Allow registering WebAuthn authenticators in memory

benthorner opened the pull request:
https://www.pivotaltracker.com/story/show/178066084

This adds basic functionality and infrastructure for registering
WebAuthn authenticators, using an in-memory store to keep the size
of the PR down. We can build on the patterns established here in
future PRs, and add more features:

- Hooking up with the API app
- Adding a graphic when there are no keys
- Managing existing keys
- Dealing with server-side errors
- Authentication

**This PR is complicated.** I've tried to be as helpful as possible
in the code and commit messages, so have a look at those if you're
not sure about something. Most of the complexity is in the second
commit, where all the pieces can be seen working together.

## Screenshots

| No key (need to add graphic) | After adding a key |
| ------------- | ------------- |
| <img width="851" alt="Screenshot 2021-05-11 at 14 56 12" src="https://user-images.githubusercontent.com/9029009/117827646-13e4bd00-b269-11eb-9669-9ef0d280239f.png">  | <img width="843" alt="Screenshot 2021-05-11 at 14 56 24" src="https://user-images.githubusercontent.com/9029009/117827664-19da9e00-b269-11eb-8361-6888ee4dd245.png"> |

### Chrome

| New key | Existing key (also alert afterwards) |
| ------------- | ------------- |
| <img width="453" alt="Screenshot 2021-05-11 at 14 58 16" src="https://user-images.githubusercontent.com/9029009/117827966-560dfe80-b269-11eb-9679-48b9398612b3.png"> |  <img width="453" alt="Screenshot 2021-05-11 at 14 57 56" src="https://user-images.githubusercontent.com/9029009/117827994-5b6b4900-b269-11eb-898e-908c296bee2b.png"> |

### Safari

| New key | Existing key (silently errors, so alert) |
| ------------- | ------------- |
| <img width="486" alt="Screenshot 2021-05-11 at 14 59 42" src="https://user-images.githubusercontent.com/9029009/117828240-8f466e80-b269-11eb-930b-246c23c7af3d.png"> | <img width="471" alt="Screenshot 2021-05-11 at 14 59 51" src="https://user-images.githubusercontent.com/9029009/117828270-940b2280-b269-11eb-9a24-6e3115672783.png"> |

### Firefox

| New key | Existing key (silently errors, so alert) |
| ------------- | ------------- |
| <img width="375" alt="Screenshot 2021-05-11 at 15 35 24" src="https://user-images.githubusercontent.com/9029009/117833850-97ed7380-b26e-11eb-8e3d-a7d29fed2ed7.png"> | <img width="444" alt="Screenshot 2021-05-11 at 15 35 38" src="https://user-images.githubusercontent.com/9029009/117833882-9de35480-b26e-11eb-85d3-38e50b3af59b.png"> |



### MS Edge

| New key | Existing key (TODO) |
| ------------- | ------------- |
| <img width="568" alt="edge_yubikey_1st_prompt" src="https://user-images.githubusercontent.com/87140/117981761-435cfd80-b32d-11eb-8864-fe82384bea9c.png"> | <img width="556" alt="edge_yubikey_error" src="https://user-images.githubusercontent.com/87140/117995504-48c04500-b339-11eb-83a9-810abb87488a.png"> |

CrystalPea commented on app/webauthn_server.py:
```diff
@@ -0,0 +1,32 @@
+
+    # some browsers don't seem to have a default timeout
+    # 10 seconds seems like a reasonable amount of time
+    app.webauthn_server.timeout = 10_000
```
when in the flow this timeout could happen?

If it's the time that user has to touch the key / make a decision, we could make it a bit longer, maybe 30 seconds? Otherwise it's ok 👍🏼

benthorner replied:
Yeah, this is [the timeout for the presence check](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialCreationOptions/timeout).

Keeping it short makes testing the timeout easier, but we can always change it locally for testing. I originally thought 10 seconds would be enough, even for someone with a motor or visual impairment. Maybe not though if it's the first time they're seeing the dialogue. I've amended the second commit to use 30 seconds 👍.

CrystalPea commented on tests/app/main/views/test_webauthn_credentials.py:
```diff
@@ -0,0 +1,109 @@
+        values={'ADMIN_BASE_URL': 'http://localhost:6012'}
+    )
+
+    # override base URL so it's consistent on CI and locally
```
should this comment go here, or above the mocking of the base URL?

benthorner replied:
Sort of, it applies to both: `init_app` needs to be called in order to pick up the change.

I've amended the second commit to move the comment and group the code together a bit more.

CrystalPea commented on tests/app/models/test_webauthn_credential.py:
```diff
@@ -0,0 +1,39 @@
+    # had to use the cbor2 library to re-encode the attestationObject due to implementation differences
+    registration_response = {
+       'clientDataJSON': base64.b64decode(b'eyJ0eXBlIjogIndlYmF1dGhuLmNyZWF0ZSIsICJjbGllbnRFeHRlbnNpb25zIjoge30sICJjaGFsbGVuZ2UiOiAiYlB6cFgzaEhRdHNwOWV2eUtZa2FadFZjOVVOMDdQVWRKMjJ2WlVkRHA5NCIsICJvcmlnaW4iOiAiaHR0cHM6Ly93ZWJhdXRobi5pbyJ9'),  # noqa
+       'attestationObject': base64.b64decode(b'o2NmbXRoZmlkby11MmZnYXR0U3RtdKJjc2lnWEgwRgIhAI1qbvWibQos/t3zsTU05IXw1Ek3SDApATok09uc4UBwAiEAv0fB/lgb5Ot3zJ691Vje6iQLAtLhJDiA8zDxaGjcE3hjeDVjgVkCUzCCAk8wggE3oAMCAQICBDxoKU0wDQYJKoZIhvcNAQELBQAwLjEsMCoGA1UEAxMjWXViaWNvIFUyRiBSb290IENBIFNlcmlhbCA0NTcyMDA2MzEwIBcNMTQwODAxMDAwMDAwWhgPMjA1MDA5MDQwMDAwMDBaMDExLzAtBgNVBAMMJll1YmljbyBVMkYgRUUgU2VyaWFsIDIzOTI1 ...
```
do we want to have this registration_response defined here, or would we want to hide it away since it's this long encoded string?

CrystalPea replied:
Also do we want to present a decoded, readable version, to be able to visually compare with assertions?

benthorner replied:
The decoded version wouldn't be very useful, I'm afraid, as it's also CBOR-encoded. Even the CBOR-decoded version is pretty unreadable: it is a dict, but most of the values are still long byte sequences.

> would we want to hide it away

I was wondering if we had a place to put this sort of thing. It's basically a file fixture. I don't think it belongs in `conftest.py`, as it's only used once. On the other hand, I think it's OK in this file too.

Where do you think it could go instead?

CrystalPea replied:
Hmmm you are right we don't have an optimal place to put it. Maybe at the top of the file as a constant? I think we have put some bulky things at top of file like that in other places.

benthorner replied:
Sure, I've amended the second commit with the data in constant at the top. I realised the `clientDataJSON` is quite readable when decoded, so I've swapped it for that. I've moved the session data up there two, since the three pieces of data need to match / go together.

I guess having it at the top is easier to read if you have wrapping turned on in your editor.

How does it look now?

CrystalPea replied:
Thank you, looks good 👍🏼

CrystalPea commented on tests/app/test_webauthn_server.py:
```diff
@@ -0,0 +1,35 @@
+    assert app_.webauthn_server._verify('fake-domain') == allowed
+
+
+def test_server_rp_id(
```
what is `rp`? Maybe we could have a bit more descriptive test name here?

benthorner replied:
`rp` stands for "relying party", which isn't very intuitive but unfortunately that's the terminology used by the standard. Most of the WebAuthn docs I've seen use "rp" instead of "relying party". Since it relates to `rp.id` in the code, I don't think we should stray from this name; we just need to be familiar with WebAuthn. 

I've amended the second commit to expand it to `_relying_party_id(` in the tests.

CrystalPea replied:
I think that expansion helps, because we have it as "relying_party" elsewhere in the code, thanks!

CrystalPea reviewed (commented):
Great work! 💖 

I left a few comments. Some are suggestions, some are just questions ☺️ 

Do we want to integrate with API in this PR? I think we could?


Also, H1 header seems to be a bit low here:
<img width="1045" alt="Screenshot 2021-05-12 at 11 11 55" src="https://user-images.githubusercontent.com/20957548/117969875-e3138f00-b31f-11eb-9b0c-d87aa07aae94.png">

benthorner commented on app/templates/views/user-profile/security-keys.html:
```diff
@@ -0,0 +1,58 @@
+      {% else %}
+
+        <p class="govuk-body">
+          Security keys are an alternative way of signing in to Notify.
```
I'm not sure, I just used the content from @quis. 

Since this is only for Platform Admins, I think we're OK without a content review for now.

CrystalPea replied:
ok 👍🏼

karlchillmaid replied:
Agree. I'll review another time

benthorner commented on app/assets/javascripts/registerSecurityKey.js:
```diff
@@ -3,12 +3,67 @@
+                console.error(error);
+                // not all browsers show an error dialogue, so to be safe
+                // we manually pop one open here (to be improved in future!)
+                alert('Error communicating with device.\n\n' + error.message);
```
No, this is all happening on the client side. We'd need to do another server request to get it into our logs, which would be unusual for JS code. While it's not ideal for remote debugging, we do log the full error object in the console and the alert, so we can always ask the user for it (and for now we're the only users anyway).

benthorner commented on app/assets/javascripts/registerSecurityKey.js:
```diff
@@ -3,12 +3,67 @@
+      url: '/webauthn/register',
+      xhr: () => xhrOverride,
+      dataType: 'x-binary',
+      converters: { '* x-binary': (value) => value }
```
I did mention this in the commit message - it's [this StackOverflow article](https://stackoverflow.com/questions/12394622/does-jquery-ajax-or-load-allow-for-responsetype-arraybuffer) - but the comment is a bit hidden 😬.

- `dataType: 'x-binary'` is overriding jQuery's auto-detection.
- `converters: ...` is telling jQuery what to do with that data type.

We could probably use `'* text': ...` without the `dataType`, as I suspect that's what jQuery infers it as. I think it's better to be explicit, as the default jQuery behaviour is so "magical" (and not in a good way).

CrystalPea replied:
thanks!

benthorner commented on app/main/views/webauthn_credentials.py:
```diff
@@ -0,0 +1,43 @@
+            "displayName": current_user.name,
+        },
+        credentials=[],  # TODO: get from user
+        user_verification="discouraged",
```
This is referring to [one of the authenticatorSelection settings](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialCreationOptions/authenticatorSelection).

It means we don't require the user to setup a PIN, or other form of verification, on their authenticator. I've amended the second commit with a quick commit to hint at the meaning.

benthorner commented on tests/app/main/views/test_user_profile.py:
```diff
@@ -11,22 +11,35 @@
+    mocker.patch('app.user_api_client.get_webauthn_credentials_for_user', return_value=[])
+    page = client_request.get('main.user_profile')
+    security_keys_row = page.select_one('#security-keys')
+    assert ' '.join(security_keys_row.text.split()) == 'Security keys 0 Change'
```
Sure, that's fair. I've amended the first commit for test for no keys and one key in the same test.

benthorner commented:
Thanks @CrystalPea! I've responded to all your comments; just a few left where I think there might be more to discuss.

> Do we want to integrate with API in this PR? I think we could?

We certainly could, but I decided not to because:

- I wanted to get this up for review quickly, and integrating the API will take more time.
- Integrating with that PR kind of blocks this one, or at least makes it more complicated to try out.
- We should be able to do the integration in parallel with other work, but only once this PR is merged.

> Also, H1 header seems to be a bit low here

I see what you mean. However, I checked against the other settings pages, and it's consistent with those. The one page it's _not_ consistent with is the platform admin one, where the heading is closer to the top because it uses [a different technique](https://github.com/alphagov/notifications-admin/blob/78824f54fdf7288172387f8ed3b332d26d74e083/app/templates/views/user-profile/disable-platform-admin-view.html#L22) to set the heading. We seem to have at least three different ways of putting a heading on a page in the Admin app, which is just bonkers! So I think we should leave it as-is, but maybe fix the "disable platform admin" page to be consistent.

tombye commented on app/assets/javascripts/registerSecurityKey.js:
```diff
@@ -0,0 +1,69 @@
+  };
+
+  function fetchWebAuthnCreateOptions() {
+    var xhrOverride = new XMLHttpRequest();
```
Looking at the support for [window.fetch]() and the [web auth API](https://caniuse.com/webauthn), we could assume fetch is available if web auth is. You mentioned in your commit adding this code that fetch would be preferred so this might be an opportunity to use it instead.

In my opinion, it would be useful to:
- not have to have a hack to make `$.ajax` work
- not use a 'promise-like' thing like `$.ajax` with actual Promises if we can help it

benthorner replied:
Definitely, great idea! It's quite a big change, so I've added a couple more commits to switch to `fetch` (the 4th and 5th commits). How does it look now?

tombye replied:
Looks great, and the code and tests seem to describe the various scenarios a lot better I think 👍🏻

tombye commented on app/assets/javascripts/registerSecurityKey.js:
```diff
@@ -0,0 +1,69 @@
+      },
+      processData: false,
+      contentType: 'application/cbor',
+      data: CBOR.encode({
```
Would be nice to use `window.CBOR` here, to match your use on line 12
```suggestion
      data: window.CBOR.encode({
```

benthorner replied:
Sure, I've amended the second commit to do that 👍.

tombye commented on tests/javascripts/registerSecurityKey.test.js:
```diff
@@ -0,0 +1,146 @@
+  test('creates a new credential and reloads', (done) => {
+    // pretend window.navigator.credentials exists in test env
+    // defineProperty is used as window.navigator is read-only
+    Object.defineProperty(window.navigator, 'credentials', {
```
This and the other blocks that redefine DOM APIs should really have their original values reset after the test that uses them. Just to avoid any unexpected side effects in tests that run after these.

benthorner replied:
That makes sense. It kind of slipped my mind (/we don't seem to do it [in other tests](https://github.com/alphagov/notifications-admin/blob/78824f54fdf7288172387f8ed3b332d26d74e083/tests/javascripts/enhancedTextbox.test.js#L282)). In the 4th commit and onwards, I've shifted the test code to restore original values (if there were any).

How does it look now?

tombye replied:
That test without them you spotted should also have them. I'll put a fix up soon but do say if you know of any others.

This looks good now.

benthorner replied:
Thanks!

> That test without them you spotted should also have them.

I think it's done already: that test [also calls `teardown.js`](https://github.com/alphagov/notifications-admin/blob/78824f54fdf7288172387f8ed3b332d26d74e083/tests/javascripts/enhancedTextbox.test.js#L8), which is where I added the hook.

CrystalPea commented:
> I wanted to get this up for review quickly, and integrating the API will take more time.
Integrating with that PR kind of blocks this one, or at least makes it more complicated to try out.
We should be able to do the integration in parallel with other work, but only once this PR is merged.
Also, H1 header seems to be a bit low here

Fair, we can do it separately that's true.

> I see what you mean. However, I checked against the other settings pages, and it's consistent with those. The one page it's not consistent with is the platform admin one, where the heading is closer to the top because it uses a different technique to set the heading. We seem to have at least three different ways of putting a heading on a page in the Admin app, which is just bonkers! So I think we should leave it as-is, but maybe fix the "disable platform admin" page to be consistent.

I think it would be up to @quis to decide which way the consistency should go, from visual standpoint, no?

tombye commented on tests/javascripts/registerSecurityKey.test.js:
```diff
@@ -0,0 +1,146 @@
+      writable: true,
+    })
+
+    jest.spyOn(window.$, 'ajax').mockImplementation((options) => {
```
This and other mocked spies should be reset after their test is run, to avoid any unexpected cases in future tests.

benthorner replied:
Yup, I've addressed this as part of https://github.com/alphagov/notifications-admin/pull/3878#discussion_r631222693.

I'll resolve this one and we can continue any further discussion there.

tombye commented on tests/javascripts/registerSecurityKey.test.js:
```diff
@@ -0,0 +1,146 @@
+    })
+
+    jest.spyOn(window, 'alert').mockImplementation((msg) => {
+      done()
```
Jest's docs have `done` as the last line of code run, after any assertions, so we should probably do the same.

See: https://jestjs.io/docs/asynchronous#callbacks

benthorner replied:
Good spot, that makes sense. I accidentally put the fix in the 4th commit, but it's not a big change so I'll leave it there and just mention it in the commit message.

tombye reviewed (changes requested):
I read most of the code, to understand what was happening, but kept my review to the frontend bits.

I've left some comments on a few things but this looks good.

The only largish thing I noticed was the lack of states for users without JS and those whose devices/browsers don't support web auth. We've chatted about how to message users in those scenarios out of this pull request so the work is planned but worth mentioning it here as a record of that.

benthorner commented:
Thanks for the review @tombye 💯. I think I've addressed all your comments.

> the lack of states for users without JS and those whose devices/browsers don't support web auth

We do indeed have a plan to address this: I've made a note to do it as part of the card for this PR. It should be a quick, best practice thing to implement, so we should just do it (vs. saying it's not MVP and then forgetting).

> I think it would be up to @quis to decide which way the consistency should go, from visual standpoint, no?

@CrystalPea I guess it is. However, I think that's out-of-scope for this PR: once we make a decision, we'll need to change at least one other HTML file to make them consistent. The new HTML file is consistent with four other HTML files, so I'm not introducing anything new here. I'm happy to make all of them consistent in a follow-up PR, so we have fewer ways of doing headings 😄. But I don't think it should block this one going in.

klssmith commented on requirements.txt:
```diff
@@ -29,12 +29,16 @@ cachetools==4.2.1
     # via flask
 colorama==0.4.3
     # via awscli
+cryptography==3.4.7
```
If we need a version of cryptography that's > 3.4 we'll need to work out how to get round the build failing on the PaaS due to not having Rust installed

benthorner replied:
Nice spot / save! I've amended the second commit to pin it below 3.4 and explained why in comments. Deploying to Preview shows it works 🎉  (with a bit of variability, since we're using an in-memory store that's not shared between the two instances running in Preview).

tombye commented on tests/javascripts/registerSecurityKey.test.js:
```diff
@@ -77,16 +77,23 @@ describe('Register security key', () => {
   })
 
-  test('alerts if fetching WebAuthn options fails', (done) => {
+  test.each([
```
This is a really nice way of describing the different error states 👍🏻

tombye commented on app/assets/javascripts/registerSecurityKey.js:
```diff
@@ -0,0 +1,56 @@
+
+  window.GOVUK.Modules.RegisterSecurityKey = function() {
+    this.start = function(component) {
+      $(component)
```
I should have mentioned this before but can we put in a guard at the top to make it return if the needed web APIs aren't available?
```suggestion
      if (!('fetch' in window) || !('credentials' in window.navigator)) { return }
      $(component)
```

benthorner replied:
We did [discuss it before](https://github.com/alphagov/notifications-admin/pull/3878#pullrequestreview-657903585), but not from a JS point of view. I'll leave it as-is for now, and make it part of the follow-up PR so it's all in one place (I've made a not of your suggested change on the card).

tombye commented:
@benthorner I'm happy all my comments have been addressed. I added [an extra one](https://github.com/alphagov/notifications-admin/pull/3878#discussion_r631712899), sorry 😬, which the chat with you about browser support/fallbacks made me realise was needed. Everything else is 💯 so thanks for all the work.

klssmith commented:
I was happy with this too. I just had one question - when Chrome gives you the "Allow this site to see your security key?" window are we saving exactly the same data when someone clicks "Skip" vs when they click "Allow"?

benthorner commented:
Thanks all, it's time merge! 🎉

> I just had one question - when Chrome gives you the "Allow this site to see your security key?" window are we saving exactly the same data when someone clicks "Skip" vs when they click "Allow"?

Good spot @klssmith! I ran a few tests locally and clicking "Skip" means (referring to [the API docs](https://developer.mozilla.org/en-US/docs/Web/API/AuthenticatorAttestationResponse)):

- `authData.attestedCredentialData.aaguid` is set to `00000000000000000000000000000000` (masking the make/model).
- `fmt` is `none`, which goes with...
- `attStmt` is `{}`

If we verified attestation in future, then this wouldn't be allowed through. And it would be a red flag if we ever audit the credentials. For now I think it's OK to request `direct` attestation, and allow people to opt-out.
